proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,138 @@
+# Feature name
+
+* Proposal: TBD
+* Author(s): [Kevin Ballard](https://github.com/kballard)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+Add 3 new `SequenceType` functions `scan(_:combine:)`, `takeWhile(_:)`, and
+`dropWhile(_:)`, with overrides as appropriate on `CollectionType`,
+`LazySequenceType`, and `LazyCollectionType`, as well as a global function
+`iterate(_:apply:)`.
+
```

Comment:
I haven't heard any suggestions yet for alternate names that I thought were any better.

`dropWhile(_:)` is patterned after `dropFirst(_:)`. Personally I'd actually be in favor of renaming them both to `skipWhile(_:)` and `skipFirst(_:)` to avoid the potential confusion around whether they're mutating, but as long as we have `dropFirst()` then `dropWhile()` seems appropriate. Or perhaps that should be `drop(while:)` now? I'm not sure.

For `takeWhile(_:)`, the closest existing method is `prefix(_:)`, but `prefixWhile(_:)` (or `prefix(while:)`) doesn't make sense as "prefix" is a noun, not a verb (well, it is a verb too, but that means something different). I suppose we could go with `prefixHaving(_:)` (or should that now be `prefix(having:)`?) for that one.

And for `scan(_:combine:)`, I really don't know what else to call it. I'm open to suggestions, but since this one doesn't have an obvious plain english name, I'm inclined to say sticking with FP terminology makes the most sense.

Same thing with `iterate(_:apply:)`, I'm not sure what else to call it.